### PR TITLE
First blog about test classloading changes

### PR DESCRIPTION
**Pull request by holly-cummins**
I've added a blog discussing the externals of the test classloading change. 

I was also hoping to discuss the internals, but  that's harder so I kept putting it off. We ought to publish something alongside 3.22, so I've run out of runway to write an epic blog. The discussion of internals should maybe be in its own blog, which gives more time to get "part ii - the terrifying internals" written and out.

### CODE REVIEW

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+title: 'The internals (and a few externals) of Quarkus test classloading have changed'
+tags: announcement testing
+synopsis: 'The way that Quarkus loads test classes has been updated. Most tests will not need to change, but here are some things to watch out for.'
+author: hcummins
+---
+
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:12]:
```suggestion
In particular, this is not affecting Quarkus production classloaders it is only changing how `@QuarkusTest` test classes are loaded during tests.
```

assuming that is a true statement its worth mentioning this is only affecting tests, not prod.

...how about devmode? is it affected ?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:13]: would be good to give a slight more context up front stating that we know this could cause some tests that relied on sideeffects of the old way to fail - we expect most due to actual bugs just not discoverable before and are looking for feedback and below blog outlines the details.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+== What you need to do
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
+One of the goals of this change was that the rewrite didn't touch any tests in our test suite, to make sure they'd all continue working without updates.
+In practice, there have been a few hiccups and we've discovered some edge cases in the broader ecosystem.
+
+=== Dropped support
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:36]: the lines before here talks about the known regressions - should this dropped support section move above or below that so there is a natural flow ?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+- *Version 2.x of the Maven Surefire plugin*. Versions below 3.x of the Maven Surefire plugin will no longer work with `@QuarkusTest`. Version 3 of the Surefire plugin was released in 2023, so version 2 is now rather old.
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause port conflicts and cross-talk on configuration values. We're hoping to have a https://github.com/quarkusio/quarkus/issues/45785[fix] for this in the next release. As a workaround, splitting conflicting tests into separate projects should fix symptoms.
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:43]: are we sure we can keep that promise of fixing that? my understanding from the conversation was that this requirement of starting during augmentation phase would be too harsh for devservices ?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
+- Increased memory footprint running tests. For suites using multiple profiles and resources, more heap or metaspace may be needed.
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:45]: why? I would have expected less usage as there is just one classloader ...or is it because before there was one shared "heavy" classloader and many junit classloaders and now there are just many "heavy" junit classloaders ?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+=== Things to watch out for
+
+- *Test order change*. As part of the rewrite, the execution order of some tests has swapped around. Of course, we all know tests should not depend on execution order if they don't set an order explicitly. However, it's easy to not notice that a test requires a certain order... until the order changes. We discovered some tests in our own suite that were sensitive to the execution order, and other people may make similar discoveries.
```
Comment (maxandersen) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:49]: just curious - what order was it before and what is it now ?

**Review by maxandersen — CHANGES_REQUESTED** [review]
really good write up - added a few clarifying suggestions but good stuff!

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause port conflicts and cross-talk on configuration values. We're hoping to have a https://github.com/quarkusio/quarkus/issues/45785[fix] for this in the next release. As a workaround, splitting conflicting tests into separate projects should fix symptoms.
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
```
[the remark this answers is not in the document]
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:44]: I have a simple, elegant, clearly correct fix for this. Maddeningly, the fix breaks gradle, which does unorthodox things in its test classloading. I haven't yet got a fix that works for both gradle and non-gradle paths, but again, fingers crossed for next release.

If people do the TCCL workaround it will be a no-op once my fix drops. So better to remove it, because of code clutter, but harmless.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 5 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:45]: The overall number of classloaders is the same, and they're basically the same size as before. The extra pressure on the heap happens because we have more of them present on the heap at the same time. It used to be we'd close one before creating the next. It's a side effect of frontloading the classloading.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 1 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:12]: Dev mode isn't changed either. It is only `@QuarkusTest` (not even other test types). I'll add a clarifying sentence to list the exclusions explicitly.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 2 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:17]: Done!

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 1 above)
Comment (maxandersen) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:12]: ok, so that devservices start in a different way is *only* for tests not devmode?

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 5 above)
Comment (maxandersen) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:44]: gotcha....but they will all get closed/collected per test class or suite? i assume they must be cleaned up as otherwise continues testing will have a hard time :)

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 6 above)
Comment (maxandersen) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:48]: okey, so the order is the same between new runs ...just not same as old runs. cool.

Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:48]: That's better, thanks!

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 1 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:12]: Technically, they start in the same way as they've always done (in the augmentation phase), they just all do it at the same time. :) 

But flippancy aside, I'm not aware of any interactions between these changes and dev mode. "Dev mode" is a bit overloaded because we use it to mean both `quarkus:dev`, which actually launches two applications, and also the "not-test" application part of that. So in `quarkus:dev` the tests will get run in a different way, but the application side should be the same. I wondered if the frontloaded dev services could conflict with the application-side-dev-service, but I don't think that's an issue, because if there was a possibility for conflict on port numbers, it would have already manifested halfway through a test run, in the old architecture.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 5 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:44]: (We collect them as soon as we're done with the test/test profile group, same as before.)

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 1 above)
Comment (cescoffier) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:12]: No, the change only affects test, not dev. However, it affects continuous testing happening in dev mode.

Comment (cescoffier) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:12]: I would write it using multiple sentences:

The internals of Quarkus test classloading have been rewritten in 3.22.
It does **not** affect production and dev modes. However, it changes how `@QuarkusTest`, `@QuarkusIntegrationTest`, and other Quarkus tests are loaded.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+tags: announcement testing
+synopsis: 'The way that Quarkus loads test classes has been updated. Most tests will not need to change, but here are some things to watch out for.'
+author: hcummins
+---
+
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+Other test extensions haven't changed, nor has dev mode and production mode classloading.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:13]: If we say it before, we can remove this sentence.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+This change should make Quarkus testing work better, and it allowed us to fix a pile of longstanding bugs.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:14]: ...and improve the integration with test frameworks such as Pact.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+However, it did introduce a few bugs we know about, and most likely also some bugs we don't yet know about.
+We're keen to get feedback from the community so that we can get fixing.
+
+== Why?
+
+In previous versions, Quarkus tests were invoked using the default JUnit classloader, and then executed in a different, Quarkus-aware, classloader.
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:23]: point or time?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+However, some use cases didn't work. Tests using advanced JUnit features like `@TestTemplate` and `@ParameterizedTest` sometimes found that the same test code might appear to run in several classloaders in a single test, or that injected dependencies weren't always available.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:24]: should we say Junit 5 features ?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+While Quarkus extensions can do all sorts of marvellous bytecode manipulation to improve the developer experience, they cannot manipulate test classes with the same freedom that they do normal application classes.
+
+Over time, test-related defects were building up that couldn't be changed without a fundamental rewrite of test classloading. The Quarkus test code itself was also growing ever-more complex as it tried to work around various JUnit edge cases. Moving test instances from one classloader to another involved serializing and deserialization, which is harder to implement on newer JVM versions with tighter class security. For example, Quarkus used to use XStream as the serialization provider, but XStream no longer works with Java 17 and higher, because of reflection restrictions in the newer JVMs.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:28]: test classloading only? Or test execution too?

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+What if, instead, Quarkus tests were simply run in the same classloader used to to load them?
+
+== What you need to do
+
+From Quarkus 3.22 onwards, this is exactly how `@QuarkusTest` classloading works.
+What do your tests need to change in order to work with the new architecture?
+Hopefully, nothing.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:36]: I would write it: 

**nothing** (hopefully)

But it reduces the level of confidence.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+One of the goals of this change was that the rewrite didn't touch any tests in our test suite, to make sure they'd all continue working without updates.
+In practice, there have been a few hiccups and we've discovered some edge cases in the broader ecosystem.
+
+=== Known regressions
+
+- *Dev services start in the discovery phase*. Quarkus Dev Services are currently started during the augmentation phase. With the new design, all augmentation happens at the beginning of the test run. If several test classes with different Dev Service configuration are augmented before any tests are run, multiple differently-configured Dev Services may be running at the same time. This can cause port conflicts and cross-talk on configuration values. We're hoping to have a https://github.com/quarkusio/quarkus/issues/45785[fix] for this in the next release. As a workaround, splitting conflicting tests into separate projects should fix symptoms.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:42]: discovery / augmentation - we might need a link to a reference guide. I actually never heard about the discovery phase.

Comment (cescoffier) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:42]: We may have to explain augmentation somewhere (in an admonition)

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+- Increased memory footprint running tests. For suites using multiple profiles and resources, more heap or metaspace may be needed.
+
+=== Things to watch out for
+
+- *Test order change*. A side effect of the rewrite is that the execution order of test can be different. Of course, we all know tests should not depend on execution order if they don't set an order explicitly. However, it's easy to not notice that a test requires a certain order... until the order changes. We discovered some tests in our own suite that were sensitive to the execution order, and other people may make similar discoveries.
+- *Test timing change*. We also discovered that the rewrite exposed some timing issues in tests. Because classloading is frontloaded at the beginning of the test run, rather than between test executions, there's less time for asynchronous operations to finish between tests. For example, there may no longer be time for external state to 'reset' before the next test starts. This might expose some heisenbugs in test suites.
+
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:50]: Both are hidden bugs in the test suites or Quarkus, so I would say that's a good thing.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+=== Dropped support
+
+- *`@TestProfile` on `@Nested` tests.* Mixing different test profiles and test resources on `@Nested` tests is no longer supported. By definition, every `@TestProfile` must get its own Quarkus application and classloader. Having multiple classloaders execute one test isn't compatible with loading the test with the classloader used to run it.
+- *Version 2.x of the Maven Surefire plugin*. Versions below 3.x of the Maven Surefire plugin will no longer work with `@QuarkusTest`. Version 3 of the Surefire plugin was released in 2023, so version 2 is now rather old.
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:54]: Do we document that in the breaking change? I know it's very old and should not be used, but it might be a problem for some users.

File: _posts/2025-04-22-test-classloading-rewrite.adoc [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+== Next steps
+
+The main work of the test classloading rewrite has been delivered in 3.22, and has unlocked a bunch of possible improvements.
+Some test defects weren't directly fixed by the main change, but the architecture is now in place to enable a fix.
+More excitingly, test-related extensions, like the Pact extensions, can now add new features to reduce test boilerplate.
+
+As always, if you spot issues or oddities, please let us know on https://quarkusio.zulipchat.com/[zulip] or https://github.com/quarkusio/quarkus/issues[raise an issue].
+
```
Comment (cescoffier) [inline _posts/2025-04-22-test-classloading-rewrite.adoc:63]: I would mention the working group somewhere :-)

**Review by cescoffier — CHANGES_REQUESTED** [review]
Looks great! I added a few comments, nothing serious.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 17 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:63]: Yes, good point! I had every intention of mentioning it, and even pasted the link into a draft, but lost it. Will do.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 15 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:50]: Agreed. We're Helping! 😁

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 14 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:42]: That's JUnit discovery (oops!). It's a problem with working with JUnit that loads of concepts get overloaded - there's JUnit extensions and Quarkus extensions, JUnit phases and Quarkus phases ... which kind of phase was clear in my head, but no one else would have any way of knowing, so I'll clarify.

I'll also add some links for augmentation.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 8 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:13]: @maxandersen suggested the sentence, to make it very clear to people who might be skimming that production mode wasn't changing.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 10 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:23]: I'd say point, because we don't judge it by a clock, we do it relative to other events in the lifecycle. But I don't have strong feelings, if you think time is clearer?

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 8 above)
Comment (cescoffier) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:13]: I will go with the last one.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 10 above)
Comment (cescoffier) [reply _posts/2025-04-22-test-classloading-rewrite.adoc:24]: point is fine - I was just wondering.

**Comment by cescoffier** [comment]
Agreed, the admonition is too prominent. It's where I would love to have "side notes" - link it is.

File: _posts/2025-04-22-test-classloading-rewrite.adoc (hunk 1 above)
Comment (holly-cummins) [reply _posts/2025-04-22-test-classloading-rewrite.adoc]: I've gone with the last one, as suggested.

**Comment by maxandersen** [comment]
mark as approved based on the issues i commented on is fixed. let you decide when to merge after admonition fixed.

**Review by cescoffier — APPROVED** [review]
Great! 

I would postpone the publication after the 3.22 platform release announcement, maybe just after it. But, no strong opinion.

**Comment by holly-cummins** [comment]
I've adjusted the date and squashed, so we should be good to merge tomorrow. I've had one zulip message reporting the dev services startup sequence change already, so we should not delay past tomorrow, I think.

**Review by cescoffier — APPROVED** [review]
